### PR TITLE
Fast builds by default

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

Currently, we run optimized builds locally, which can be slow and difficult to debug.

In this PR we disable this to allow for the default behavior, and only enable optimized builds on CI.

We also verify that builds are optimized when benchmarking.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by rkooo567** [comment]
Can you explain how to turn on runtime optimized build at the `development tips` section? https://docs.ray.io/en/master/development.html

**Comment by mehrdadn** [comment]
@rkooo567 Sure, just did!

**Comment by rkooo567** [comment]
I am totally down for this change, but I don't really have good understanding cpp compilation & bazel build, so please add other people to review this PR!!

**Comment by ericl** [comment]
How does this impact the setup py develop workflow, can we make it off by default? I'm concerned this will greatly slow down performance for a common use case. Many times dev builds have to be run with fill performance.

**Comment by mehrdadn** [comment]
@ericl Optimized builds don't appear to be the most preferred use case judging the quick poll I took. Most people preferred faster builds and better debuggability. That said, all this does is to avoid overriding whatever the current setting is. As I've [explained this in the documentation](https://github.com/ray-project/ray/pull/8925/files#diff-7d3c5ad006a8661756d0496fca112d11R64), if you want to have optimized performance all the time, you can just add `build --compilation_mode=opt` to your `~/.bazelrc`.

**Comment by ericl** [comment]
Why not make this opt in by default? Such as straw man FASTBUILD=1? Btw, a
lot of ray developers care about testing the speed of their changes, after
all performance is one of the most important priorities for Ray.

**Comment by mehrdadn** [comment]
@eric-jj Because then it conflicts with your user settings. You can just add your preference to your `~/.bazelrc` though. I don't see why that wouldn't work?

File: python/ray/_raylet.pyx
Comment (edoakes) [inline python/ray/_raylet.pyx:112]: What is this used for? Is it some magic constant for cython?

File: doc/source/development.rst
Comment (edoakes) [inline doc/source/development.rst:45]:
```suggestion
This is to ensure that compilation remain as fast as possible
```

File: doc/source/development.rst
Comment (edoakes) [inline doc/source/development.rst:49]:
```suggestion
For an optimized build, you can do the following:
```

File: doc/source/development.rst
Comment (edoakes) [inline doc/source/development.rst:56]:
```suggestion
This will build Ray's Python package with optimizations (this may take a while).
```

File: python/ray/_raylet.pyx
Comment (edoakes) [reply python/ray/_raylet.pyx:112]: I see, it's used to check if the build was optimized, nice. Can you add a comment explaining this?

Comment (edoakes) [inline python/ray/_raylet.pyx:112]:
```suggestion
OPTIMIZED = __OPTIMIZE__
```

**Review by edoakes — COMMENTED** [review]
LGTM. I think this will be a big improvement for most ray developers and the warning in `ray microbenchmark` seems like it should avoid issues with benchmarking unoptimized builds. My only concern would be if there are other benchmarks where we should insert the same checks - maybe in the release tests?

Let's wait for @ericl to sign off as well before merging.

File: python/ray/_raylet.pyx
Comment (mehrdadn) [reply python/ray/_raylet.pyx:112]: Yup sure!

**Review by ericl — DISMISSED** [review]
I still don't get why we can't make this opt in. That both preserves the status quo and allows users to get a faster build if they want right, by editing their bazelrc. You can print out the tip if building in optimized mode.

**Comment by mehrdadn** [comment]
Alright, well given it seems this isn't happening, I took out the line that changes the behavior change. Shall we merge?

File: doc/source/development.rst
Comment (ericl) [inline doc/source/development.rst:44]: The doc needs to be flipped now right?

Comment (mehrdadn) [reply doc/source/development.rst]: Yeah I guess, I updated it.

**Review by ericl — APPROVED** [review]

File: doc/source/development.rst
Comment (simon-mo) [inline doc/source/development.rst:42]:
```suggestion
Fast, Debug, and Optimized Builds
---------------------------------
```

**Review by simon-mo — APPROVED** [review]

**Pull request merged**
